compiler/rustc_fs_util/src/lib.rs

```diff
@@ -102,3 +104,45 @@ pub fn path_to_c_string(p: &Path) -> CString {
 pub fn try_canonicalize<P: AsRef<Path>>(path: P) -> io::Result<PathBuf> {
     fs::canonicalize(&path).or_else(|_| absolute(&path))
 }
+
+pub struct TempDirBuilder<'a, 'b> {
+    builder: tempfile::Builder<'a, 'b>,
+}
+
+impl<'a, 'b> TempDirBuilder<'a, 'b> {
+    pub fn new() -> Self {
+        let mut builder = tempfile::Builder::new();
+        builder.rand_bytes(8);
+        Self { builder }
+    }
+
+    pub fn prefix<S: AsRef<OsStr> + ?Sized>(&mut self, prefix: &'a S) -> &mut Self {
+        self.builder.prefix(prefix);
+        self
+    }
+
+    pub fn suffix<S: AsRef<OsStr> + ?Sized>(&mut self, suffix: &'b S) -> &mut Self {
+        self.builder.suffix(suffix);
+        self
+    }
+
+    pub fn tempdir_in<P: AsRef<Path>>(&self, dir: P) -> io::Result<TempDir> {
+        let dir = dir.as_ref();
+        // On Windows in CI, we had been getting fairly frequent "Access is denied"
+        // errors when creating temporary directories.
+        // So this implements a simple retry with backoff loop.
+        #[cfg(windows)]
+        for wait in 1..11 {
+            match self.builder.tempdir_in(dir) {
+                Err(e) if e.kind() == std::io::ErrorKind::PermissionDenied => {}
+                t => return t,
+            }
+            std::thread::sleep(std::time::Duration::from_millis(1 << wait));
+        }
+        self.builder.tempdir_in(dir)
+    }
+
+    pub fn tempdir(&self) -> io::Result<TempDir> {
+        self.builder.tempdir_in(std::env::temp_dir())
```

Comment:
```suggestion
        self.tempdir_in(std::env::temp_dir())
```
I guess this ^^^ was the intention (to enable retries).
Otherwise this could just use `self.builder.tempdir()`.

Reply:
Yes, this is intentional. And eventually I would like to remove `tempdir` entirely so everything is placed in the build directory (useful for `-C save-temps`).

Reply:
>Yes, this is intentional. 

What exactly - not retrying, not using the equivalent `self.builder.tempdir()`, or both?

Reply:
Oh wait, derp. This was unintentional. I did indeed mean `self.tempdir_in`. Somehow my mind kept missing that.